modin/engines/base/io/text/text_file_dispatcher.py

```diff
@@ -153,56 +157,123 @@ def partitioned_file(
             Indicate quote in a file.
         is_quoting: bool, default True
             Whether or not to consider quotes.
+        header_size: int, default 0
+            Number of rows, that occupied by header.
+        pre_reading: int, default 0
+            Number of rows between header and skipped rows, that should be read.
+        encoding: str, default None
+            `encoding` parameter of read_* function.
 
         Returns
         -------
-        An array, where each element of array is a tuple of two ints:
-        beginning and the end offsets of the current chunk.
+        List of tuples with the next elements:
+            int: partition start read byte
+            int: partition end read byte
+            int, array-like or callable: skiprows object aligned (adopted)
+                to the exact partition
+            int: number of rows that exact partition will contain
         """
+        read_rows_counter = 0
+        considered_rows_counter = 0
+        outside_quotes = True
+        if encoding is not None:
+            partition_skiprows = 1
+            # if encoding is not None, parser will read additional line
+            # with header
+            read_rows_counter -= 1
+        else:
+            partition_skiprows = None
+        should_handle_skiprows = skiprows is not None and not isinstance(skiprows, int)
         if num_partitions is None:
-            num_partitions = NPartitions.get()
+            num_partitions = NPartitions.get() - 1 if pre_reading else NPartitions.get()
```

Comment:
Why does `pre_reading` affect `NPartitions`?

Reply:
If `pre_reading` is not None, the first `pre_reading` lines will be read in the first partition and the rest of the file should be splitted into `NPartitions.get() - 1` partitions.

Reply:
It turns out that we may have a case when in the first part there are very few lines, and in the rest there are a lot, right?

Reply:
Yes, you are correct, it's a bad corner case, but in that case we use pre_reading, that occurs by "squashing" complex `skiprows` into integer, that we can process easily and avoid the overhead caused by handling of complex `skiprows` case.